sw/device/lib/dif/dif_rv_timer.c

```diff
@@ -0,0 +1,194 @@
+}
+
+dif_rv_timer_get_tick_params_result_t dif_rv_timer_get_tick_params(
+    const dif_rv_timer_t *timer, dif_rv_timer_tick_params_t *params_out) {
+  if (timer == NULL) {
+    return kDifRvTimerGetTickParamsBadArg;
+  }
+
+  uint32_t cfg_value = mmio_region_read32(
+      timer->base_addr, offset_by_hart(RV_TIMER_CFG0_REG_OFFSET, timer->hart));
+
+  params_out->prescale =
+      (uint16_t)((cfg_value >> RV_TIMER_CFG0_PRESCALE_OFFSET) &
+                 RV_TIMER_CFG0_PRESCALE_MASK);
+  params_out->tick_step = (uint8_t)((cfg_value >> RV_TIMER_CFG0_STEP_OFFSET) &
+                                    RV_TIMER_CFG0_STEP_MASK);
+
+  return kDifRvTimerGetTickParamsOk;
+}
+
+dif_rv_timer_counter_read_result_t dif_rv_timer_counter_read(
+    const dif_rv_timer_t *timer, uint64_t *out) {
+  if (timer == NULL) {
+    return kDifRvTimerCounterReadBadArg;
+  }
+
+  while (true) {
+    uint32_t upper = mmio_region_read32(
+        timer->base_addr,
+        offset_by_hart(RV_TIMER_TIMER_V_UPPER0_REG_OFFSET, timer->hart));
+    uint32_t lower = mmio_region_read32(
+        timer->base_addr,
+        offset_by_hart(RV_TIMER_TIMER_V_LOWER0_REG_OFFSET, timer->hart));
+
+    /// Read the upper half again to make sure the lower
+    /// half didn't overflow right after we read it.
+    uint32_t overflow_check = mmio_region_read32(
+        timer->base_addr,
+        offset_by_hart(RV_TIMER_TIMER_V_UPPER0_REG_OFFSET, timer->hart));
+
+    if (upper == overflow_check) {
+      *out = (((uint64_t)upper) << 32) | lower;
+      return kDifRvTimerCounterReadOk;
+    }
+  }
+}
+
+dif_rv_timer_arm_result_t dif_rv_timer_arm(const dif_rv_timer_t *timer,
+                                           uint64_t threshold) {
+  if (timer == NULL) {
+    return kDifRvTimerArmBadArg;
+  }
+
+  /// At least higher than the current value; cannot set off the alarm.
+  mmio_region_write32(
+      timer->base_addr,
+      offset_by_hart(RV_TIMER_COMPARE_UPPER0_0_REG_OFFSET, timer->hart),
+      UINT32_MAX);
+  /// At least higher than the new value; if this happens to trigger an
+  /// interrupt, the following wright would have done so, anyway.
+  mmio_region_write32(
+      timer->base_addr,
+      offset_by_hart(RV_TIMER_COMPARE_LOWER0_0_REG_OFFSET, timer->hart),
+      (uint32_t)threshold);
+  /// Finish writing the new value; this may set off an alarm immediately.
+  mmio_region_write32(
+      timer->base_addr,
+      offset_by_hart(RV_TIMER_COMPARE_UPPER0_0_REG_OFFSET, timer->hart),
+      (uint32_t)(threshold >> 32));
+
+  return kDifRvTimerArmOk;
+}
+
+dif_rv_timer_set_enabled_result_t dif_rv_timer_set_enabled(
+    const dif_rv_timer_t *timer, dif_rv_timer_enabled_t enabled) {
+  if (timer == NULL) {
+    return kDifRvTimerSetEnabledBadArg;
+  }
+
+  switch (enabled) {
+    case kDifRvTimerEnabled:
+      mmio_region_nonatomic_set_bit32(timer->base_addr,
+                                      RV_TIMER_CTRL_REG_OFFSET, timer->hart);
+      return kDifRvTimerSetEnabledOk;
```

Comment:
Nit: both of success cases return the same error code, could just break and return outside of the switch.

Reply:
I think local returns are more readable.